src/util/simplify_expr_int.cpp

```diff
@@ -848,32 +856,43 @@ bool simplify_exprt::simplify_bitwise(exprt &expr)
 
 bool simplify_exprt::simplify_extractbit(exprt &expr)
 {
-  const typet &op0_type=expr.op0().type();
-
-  if(!is_bitvector_type(op0_type))
-    return true;
+  PRECONDITION(expr.id() == ID_extractbit);
+  auto &extractbit_expr = to_extractbit_expr(expr);
 
-  std::size_t width=to_bitvector_type(op0_type).get_width();
+  const typet &src_type = extractbit_expr.src().type();
 
-  assert(expr.operands().size()==2);
+  if(!is_bitvector_type(src_type))
+    return true;
 
-  mp_integer i;
+  std::size_t src_bit_width = to_bitvector_type(src_type).get_width();
 
-  if(to_integer(expr.op1(), i))
+  if(extractbit_expr.index().id() != ID_constant)
+  {
     return true;
+  }
 
-  if(!expr.op0().is_constant())
+  auto index_converted_to_int =
+    numeric_cast<mp_integer>(extractbit_expr.index());
+  if(!index_converted_to_int.has_value())
+  {
+    return true;
+  }
+  mp_integer index_as_int = index_converted_to_int.has_value();
+  if(!extractbit_expr.src().is_constant())
     return true;
 
-  if(i<0 || i>=width)
+  if(index_as_int < 0 || index_as_int >= src_bit_width)
     return true;
 
-  const irep_idt &value=expr.op0().get(ID_value);
+  const irep_idt &src_value =
+    to_constant_expr(extractbit_expr.src()).get_value();
 
-  if(value.size()!=width)
+  if(src_value.size() != src_bit_width)
     return true;
 
-  bool bit=(id2string(value)[width-integer2size_t(i)-1]=='1');
+  bool bit =
+    (id2string(src_value)
```

Comment:
Do `id2string` earlier on as doing `src_value.size()` will also require building the string.

Reply:
Done.